### ISSUE TITLE
Annotation parsing incorrect when class scope defines same name as import

### ISSUE THREAD

**Issue by PeterJCLaw** [python/mypy#12414]
**Bug Report**

Given:
``` python
import uuid
class Bar:
    uuid: uuid.UUID = uuid.uuid4()
reveal_type(Bar().uuid)
```

`mypy` reports no errors and will inform the user that the type of the attribute is `uuid.UUID`.

However this isn't strictly correct and without lazy annotations enabled this code will actually error at runtime:
```
  File ".../demo.py", line 3, in Bar
    uuid: uuid.UUID = uuid.uuid4()
AttributeError: 'UUID' object has no attribute 'UUID'
```

Ideally mypy would report the missing attribute (`UUID`) along with a mention of the name overlap between the type name and the attribute name.

Cross-linking: a related issue was raised against pyflakes (that the import appeared unused; https://github.com/PyCQA/pyflakes/issues/682; https://github.com/PyCQA/pyflakes/issues/648), though rejected on the grounds that this code is indeed invalid. Unfortunately when this code is alongside other code it is easy for there not to be any linter errors.

**Your Environment**

- Mypy version used: 0.910, also tested 0.941
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none that matter
- Python version used: 3.9
- Operating system and version: Ubuntu 20.04

**Comment by erictraut** [comment]
This is a subtle issue. The problem occurs only when the following conditions exist:
1. The class variable assignment includes a type annotation
2. Type annotation evaluation is not deferred
3. The expression on the RHS of the assignment references a symbol with the same name as the class variable

It appears that the Python interpreter adds the class variable to the symbol table at different times depending on whether a type annotation is present (and its evaluation is not deferred). If the type annotation is present, the variable's symbol is added to the symbol table prior to evaluating the RHS expression. In other cases, it's added only after evaluating the RHS expression. I'm not aware of this inconsistent behavior being documented anywhere, so I'm not sure if it's intentional or a bug in the CPython interpreter. To address this bug, the mypy binder will need to model this subtle behavior.
